Implement new Kibana query language

Bargs opened the pull request:
Replaces https://github.com/elastic/kibana/pull/11915 (changes from that PR are in the first commit)
Fixes https://github.com/elastic/kibana/issues/12282
Part of https://github.com/elastic/kibana/issues/10789

See the [issue description](https://github.com/elastic/kibana/issues/12282) for an up to date reference for the new syntax. This PR has everything in it that I think is necessary for 6.0 beta1. 

To enable query language switching, go to Advanced Settings and enable `search:queryLanguage:switcher:enable`. You can also set kuery as the default language by updating `search:queryLanguage`.

Some caveats: 
* Doesn’t support filters created via filters agg. This would be a nice to have, but since we only just started supporting filters from the filter agg *at all*, I don't see it as a blocker. Need some time to think about how we should do it. I don't really want to add support for query_string_query and raw query DSL natively in kuery. Including them as aliased sub-queries might be one option.
* There's no way to migrate saved searches written in lucene.
* Geo queries can get quite long very quickly. It's pretty obvious we'll need multi-line support. I'd like to explore this in another PR in the near future.

TODO:
- [x] Add a crapload of tests

Release Note: This PR adds a new experimental query language to Kibana. The new language is turned off by default but can be enabled in the Management > Advanced Settings via the `search:queryLanguage:switcher:enable` option. Changing this setting to true will add a language selector to the query bar that allows users to choose the new language for their query. Details of the new query language can be found in the issue description (https://github.com/elastic/kibana/issues/12282).

![screen shot 2017-07-13 at 3 17 21 pm](https://user-images.githubusercontent.com/6239176/28684423-56300546-72d2-11e7-8dfb-00214d01cd03.png)

Bargs commented on src/core_plugins/kibana/public/dashboard/dashboard.js:
```diff
@@ -228,6 +241,27 @@ app.directive('dashboardApp', function ($injector) {
+        $scope.refresh();
+      });
+
+      $scope.$watch(() => dashboardState.getAppState().$newFilters, function (filters = []) {
```
I moved all the filter handling logic into the `query_manager` module, but there's still a bit of tech debt here with this watch. I've got another branch where I'm working on removing the need for a watch, but I don't think it's necessary for this PR.

weltenwort commented:
After playing around with it a bit I have to say it already feels more solid that the lucene query syntax :wink: Here are a few observations though:

* Maybe being more lenient with whitespace would be good: `( range:[1 to 10] )` is allowed but `(range:[ 1 to 10 ])` or `range:[1  to  10]` is not
* At first I was confused that `(is(extension, jpg)) (and) (is(response, 200))` is valid until I realized it is interpreted as `is(extension, jpg) and is(<ANYFIELD>, and) and is(response, 200)`. Maybe we should provide some kind of feedback to the user to make that clear. I was thinking of maybe automatically reformatting the query to a canonical, unambiguous form. But maybe that is too invasive?
* It was surprising to notice that `*:"win xp"` and `machine.os:"win xp"` are interpreted differently (the former is a `query_string` query with all its syntax and the latter a `match_phrase` query). This allows to escape kuery using quotes: `"_exists_:host"`. Is that on purpose?
* To me the range query shorthand `a:[b TO c]` feels like it should be inclusive, i.e. equivalent to `range(a, gte=b, lte=c)`.

Overall it feels quite intuitive so far. I will dig into the code next.

Bargs commented:
Good points @weltenwort, you've touched on some important and thorny issues.

> At first I was confused that (is(extension, jpg)) (and) (is(response, 200)) is valid until I realized it is interpreted as is(extension, jpg) and is(<ANYFIELD>, and) and is(response, 200). Maybe we should provide some kind of feedback to the user to make that clear. I was thinking of maybe automatically reformatting the query to a canonical, unambiguous form. But maybe that is too invasive?

This is a tricky subject for sure. I'm a bit wary of automatic reformatting. As a user I'm always a bit surprised and confused when an app does this to me. An option to manually kick off an auto-reformat would be cool, but maybe that's a feature for another PR.

I think the simplest option for now would be to consider `and` and `or` keywords of the query language. The example you provided would then throw a parse error. If the user actually wanted to query for the word "and" or "or", they would need to wrap it in double quotes. How does that sound?

The nuclear option would be to require double quotes around all search terms, as you mentioned in the last PR. I'm not quite ready to pull that trigger yet since no-one likes typing double quotes constantly. 

> It was surprising to notice that *:"win xp" and machine.os:"win xp" are interpreted differently (the former is a query_string query with all its syntax and the latter a match_phrase query). This allows to escape kuery using quotes: "_exists_:host". Is that on purpose?

Not on purpose :/   I obviously didn't think this one all the way through. I'm just abusing the `query_string_query` for its all_fields mode. I initially tried using the `multi_match` query with `"fields": ["*"]` but got the error `The _source field is not searchable`, so I switched to the query_string_query. Any idea how to execute a match query across all fields? My research is coming up nil, so I may need to reach out to the ES team on this one. Perhaps @dakrone could impart some wisdom.

> To me the range query shorthand a:[b TO c] feels like it should be inclusive, i.e. equivalent to range(a, gte=b, lte=c)

Me too, this is a bug :)   good catch!

Bargs commented:
@weltenwort I've addressed everything except for the query_sting_query issue

weltenwort commented:
Regarding the phrase search across all fields: I imagine we could use a `query_string_query` and wrap it in quotes (`\`-escaping user-supplied quotes) to force a phrase search. I cannot judge whether there are other caveats with that approach though.

dakrone commented:
> I initially tried using the multi_match query with "fields": ["*"] but got the error The _source field is not searchable, so I switched to the query_string_query. Any idea how to execute a match query across all fields? My research is coming up nil, so I may need to reach out to the ES team on this one. Perhaps @dakrone could impart some wisdom.

I am familiar with this, I tried this initially when working on the `all_fields` mode with the same problem, the issue is that a blanket wildcard includes some fields that are not (and should not) be searchable.

Adrien opened an issue to allow `fields:*` queries https://github.com/elastic/elasticsearch/issues/25551, which internally can rewrite to the correct combination of fields (we do this now with `all_fields` by having a list of "valid" fields and field-types that should be included in the field list)

Bargs commented:
@dakrone thanks for the info, you guys are great!

@weltenwort I'll try your suggestion as a temporary measure, I think it should work alright until we can do a proper match query across all fields.

Bargs commented:
Note: function syntax (e.g. `is(response, 200)`) is slightly broken at the moment, I'm in the middle of extracting the AST node creation code out of the grammar file and writing grammar tests. I'll get this finished up tomorrow.

Bargs commented on src/ui/public/kuery/lib/functions/not.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+export function toKueryExpression(node) {
+  if (node.serializeStyle !== 'operator') {
```
It's a [unary operator](https://en.wikipedia.org/wiki/Unary_operation) in prefix notation.

weltenwort replied:
You could make the same argument for `:`.

Bargs replied:
Yes, that's exactly the argument I made about `:` above :)   I wonder if you're looking at an outdated version of the PR? I think at one point I categorized `:` as `shorthand`, but I've since updated it to `operator`.

I'll try to sum up my thoughts here since I can't do an inline reply to your review summary below. I don't believe I've used these terms inconsistently. I think we're just operating with different definitions in mind. Here's how I categorize things:

* `implicit` (just `and`)
* `operator` (`foo:bar`, `foo or bar`, `foo and bar`, `-foo`) (basically anything the grammar has a special case for)
* `function` (`is(foo, bar)`)
* `shorthand` (just range `bytes:[1000 to 8000]`)

I wouldn't be surprised if these categorizations evolved along with the language, but they seem to work ok for now.

weltenwort replied:
Yes, just having the definitions written down does a lot for the understanding. :+1:

I still don't get why `:` and `:[ to ]` are different categories, but maybe future additions will make that distinction clearer to me.

Bargs replied:
`:[ to ]` just seemed weird to me... it felt wrong to call it an operator since it's not strictly prefix or infix. However, after giving it some more thought and perusing the [Operator Wikipedia page](https://en.wikipedia.org/wiki/Operator_(computer_programming)), maybe it's ok. It's not much weirder than the ternary operator. I thought calling it `shorthand` instead of `operator` would reduce confusion, but it seems that's not the case ;)

Bargs replied:
4d64349cfaba19062ad0cff6feee8fa2d427a276

Bargs commented on src/ui/public/kuery/lib/functions/is.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+export function toKueryExpression(node) {
+  if (node.serializeStyle !== 'operator') {
```
`foo is bar` isn't valid. I think of `:` as a binary operator. I think my use of the term `shorthand` for the range alias `:[]` may be confusing things here. I didn't use the term `operator` there because it's a weird syntax that sort of half wraps the arguments, I don't think most people would think of it as an operator. But we could call it an operator and get rid of the `shorthand` term entirely, if you think that would be less confusing.

weltenwort replied:
I know `foo is bar` is not valid in the current implementation. But one might expect it to me. In any case I tried to delineate the different cases by giving examples.

Bargs commented on src/ui/public/kuery/lib/filter_migration/lib/__tests__/exists.js:
```diff
@@ -0,0 +1,35 @@
+      expect(result.arguments[0].value).to.be('foo');
+    });
+
+    it('should return undefined if the given filter is not of type "exists"', function () {
```
Returning undefined and throwing an exception for a non-exceptional scenario both seem equally hacky to me. Since these are private functions used only by `filterToKueryAST`, it doesn't seem terribly important, does it? 

If you're strongly against returning `undefined`, perhaps I should add a `supports` function to each converter module that returns true/false depending on whether that module supports the given filter?

weltenwort replied:
If we had a type system passing something different from an "exists" filter as an argument would be a type error. So I would absolutely call this an exceptional scenario. Returning `undefined` just has no intrinsic meaning and can be caused by all kind of accidental mistakes in JS. I would be in favor of clearly signalling the "here is your converted filter" and "I can not convert that". Adding a `supports` function vs throwing an exception essentially comes down to the "ask permission" rather than "forgiveness" pattern. I would be fine with either.

Bargs replied:
But we don't have a type system :)   an exceptional scenario in one language is another language's bread and butter. Anyway, it doesn't matter much to me since this isn't a public API. I'll just switch them to exceptions since it'll be easier than adding a new function.

Bargs replied:
4394ec061d87341799feeb1c0b984c3dae06eeaa

weltenwort replied:
thanks :smile: just want to point out that even though there is no way in javascript to write down function signatures or even enforce them, the functions still have one.

Personally, as a developer who has to reconstruct those signatures in his head on the fly, I always find `undefined`s a major pain. Every piece of code that works with the results has to check for it and whenever I return to the code I have to derive the meaning of `undefined` again. I appreciate this freeing up of cognitive capacity. :unicorn:

Bargs replied:
I suppose it's just one of those things that comes down to personal preference. I have the exact same feelings about exceptions XD

If only javascript had proper interfaces and polymorphism it would be a non-issue because we wouldn't have to write this hacky switching logic ourselves.

weltenwort commented on src/ui/public/kuery/lib/functions/is.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+function isDoubleQuoted(str) {
+  return (_.first(str) === '"') && (_.last(str) === '"');
```
This works, but using `str.startsWith('"')` might be more idiomatic.

Bargs replied:
👍  69447bfb1b25a4f5b56768d56d70517f2ff3237b

weltenwort commented on src/ui/public/kuery/lib/functions/is.js:
```diff
@@ -0,0 +1,68 @@
+      }
+    };
+  }
+  else if (fieldName !== '*' && _.isUndefined(value)) {
```
As commented earlier, I am not sure this is a good idea.

Bargs replied:
👍  b73bd2bbe40e7d308273d9d44af4f9ec0bab4b91

weltenwort commented on src/ui/public/kuery/lib/functions/not.js:
```diff
@@ -0,0 +1,38 @@
+export function toElasticsearchQuery(node, indexPattern) {
+  let [ argument ] = node.arguments;
+  if (argument.type === 'literal') {
+    argument = nodeTypes.function.buildNode('is', '*', argument.value);
```
Similar to the other case, I feel this should have happened earlier (probably in `buildNodeParams`).

Bargs replied:
See comment above

Bargs commented on src/ui/public/kuery/lib/filter_migration/index.js:
```diff
@@ -1,21 +1 @@
-
-  return negate ? nodeTypes.function.buildNode('not', node) : node;
-}
+export { filterToKueryAST } from './lib/filter_to_kuery';
```
I feel like shoving everything into the top level `public` directory makes it very difficult to understand the most important relationships between modules. For example, ask a newcomer to infer the relationships of all the vislib code scattered throughout `public` and they're gonna have a bad time.

Let me propose a couple options that cut down on the `/lib/lib/lib`ness while still grouping all the kuery code in one spot. 

Option 1

![screen shot 2017-07-10 at 11 03 50 am](https://user-images.githubusercontent.com/6239176/28024736-8a5396ba-655f-11e7-88a7-269d8095c823.png)

Option 2

![screen shot 2017-07-10 at 11 00 46 am](https://user-images.githubusercontent.com/6239176/28024597-2e1947f0-655f-11e7-8e2e-5b9355225f8c.png)

weltenwort replied:
:+1: for the second option

I agree that putting everything on the top level would make code paths too hard to trace.

Bargs replied:
👍  0a8321158496c313fa7d95d1d7e011905564c1f7

weltenwort commented on src/ui/public/kuery/lib/functions/not.js:
```diff
@@ -0,0 +1,38 @@
+  const [ argument ] = node.arguments;
+  const queryString = ast.toKueryExpression(argument);
+
+  if (argument.function && (argument.function === 'and' || argument.function === 'or')) {
```
It might be less coupled to `and` and `or` specifically if this looked at the serialization style to decide whether it is an infix that needs grouping or whether it is a "self-grouping" prefix style expression. That way it would also leave out the parentheses with prefix-style `and`s and `or`s.

Bargs replied:
I think you're right, this will need to become smarter eventually. I don't have the granularity required in `serializeStyle` for what you describe yet, and I'm not quite ready to dive into overhauling that. I think the best solution will become more clear to me over time, so I'd like to defer this until later.

Bargs replied:
Forgot to mention: I updated this condition to check for the `function` serializeStyle, so that in the meantime we won't run into the unnecessary wrapping you described.

03ed9220e80bfab126e6d4be90be809cffc13435

weltenwort commented on src/ui/public/kuery/lib/functions/range.js:
```diff
@@ -0,0 +1,79 @@
+
+export function toElasticsearchQuery(node, indexPattern) {
+  const [ fieldNameArg, ...args ] = node.arguments;
+  const fieldName = ast.toElasticsearchQuery(fieldNameArg);
```
In other places the `fieldName` is hardcoded to be a `literal`. Is this case different on purpose?

Bargs replied:
No real reason, it should probably be `literal` everywhere
ec1ae47be4465c04a6820f7dc0fc4b87b83a2884

Bargs commented on src/ui/public/filter_bar/lib/map_geo_polygon.js:
```diff
@@ -0,0 +1,21 @@
+  return function (filter) {
+    if (filter.geo_polygon) {
+      return courier
+      .indexPatterns
```
I have to object here. I'm following the existing style of all the other files in this module. Changing just this one file makes it inconsistent and harder to read in context. We either need to revert this indentation rule (which was thoughtlessly changed) or fix the entire codebase in one go, otherwise it's just making things worse. I'll open a separate issue to discuss this.

Bargs replied:
https://github.com/elastic/kibana/issues/12729

Bargs commented on src/core_plugins/kibana/public/dashboard/dashboard.js:
```diff
@@ -267,16 +255,11 @@ app.directive('dashboardApp', function ($injector) {
-        }
+        // These are handled by the filter bar directive when lucene is the query language
+        Promise.all(filters.map(queryManager.addLegacyFilter))
+        .then(updateState)
```
I've updated this spot to maintain consistency with the rest of this particular file

30029ebfdedaf1da6baa9fc240637d0bcc62d9b0

Bargs commented on src/ui/public/kuery/lib/filter_migration/lib/__tests__/filter_to_kuery.js:
```diff
@@ -0,0 +1,47 @@
+      const result = filterToKueryAST(filter);
+
+      expect(result).to.have.property('type', 'function');
+      expect(result).to.have.property('function', 'not');
```
👍  aa2c4783f215ecfb069f818eadb00fad780c88b3

Bargs commented on src/ui/public/kuery/lib/__tests__/ast.js:
```diff
@@ -1,69 +1,243 @@
+        ], 'operator')
+      ], 'operator');
+      const actual = fromKueryExpressionNoMeta('foo or bar and baz');
+      expectDeepEqual(actual, expected);
```
👍  d4bbd18cdaeb65bb5a8f7665d14359d9573e1025

weltenwort commented on src/ui/public/query_manager/index.js:
```diff
@@ -0,0 +1,79 @@
+
+  return function (state) {
+
+    function add(field, values = [], operation, index) {
```
> And put your optional arguments at the end.

-- [Kibana JavaScript Style Guide](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#default-argument-syntax)

Aside from that, maybe we should take this chance to improve this function's signature? Or is it meant to be a compatibility adapter?

Bargs replied:
The idea is that this will become the API we pass to clients who want to manipulate the query/filters. In particular, the vis team needs an API to pass to visualizations. Mostly I was being lazy here because I'm going to have to revisit and expand this API before handing it to the vis team. I figured I'd get a better idea of what the function signatures should look like once I start working on the rest of the API.

Bargs commented on src/ui/public/kuery/lib/__tests__/ast.js:
```diff
@@ -1,69 +1,243 @@
+      expect(ast.toKueryExpression()).to.be('');
+
+      const node = nodeTypes.function.buildNode('exists', 'foo');
+      delete node.type;
```
👍  7b1c4e93949115c119bee917e5d5164667dfd962

Bargs commented on src/ui/public/kuery/lib/__tests__/ast.js:
```diff
@@ -1,69 +1,243 @@
 
-      expect(parseResult).to.eql(ast);
+      const node = nodeTypes.function.buildNode('exists', 'foo');
+      delete node.type;
```
👍  7b1c4e93949115c119bee917e5d5164667dfd962

Bargs commented on src/ui/public/kuery/lib/__tests__/ast.js:
```diff
@@ -1,69 +1,243 @@
+  describe('symmetry of to/fromKueryExpression', function () {
 
-      expect(parseResult).to.eql(ast);
+    it('converting from an expression to an AST and back again should result in the same expression', function () {
```
👍  293279e264887dd4fffdea3ef47a1fe70fd779d4

Bargs commented on src/ui/public/kuery/lib/functions/__tests__/and.js:
```diff
@@ -0,0 +1,97 @@
+
+    describe('toKueryExpression', function () {
+
+      it('should serialize "and" nodes with an implicit syntax by default', function () {
```
👍  a6f80f1cffac10a654a014c6fe1d8bbe256f53ab

Bargs commented on src/ui/public/kuery/lib/functions/__tests__/is.js:
```diff
@@ -0,0 +1,128 @@
+        expect(result.simple_query_string.query).to.be('"+response"');
+      });
+
+      it('should return an ES exists query when no value is provided', function () {
```
Agreed, both arguments should be required. I think this was left over from some intermediate implementation.

b73bd2bbe40e7d308273d9d44af4f9ec0bab4b91

Bargs commented on src/ui/public/kuery/lib/functions/__tests__/range.js:
```diff
@@ -0,0 +1,109 @@
+        const result = range.buildNodeParams('bytes', givenParams);
+        const { arguments: [ , ...params ] } = result;
+
+        params.map((param) => {
```
👍  07804f28c2e9072a991132c7413758f78c7f144d

Bargs commented on src/ui/public/kuery/lib/functions/and.js:
```diff
@@ -0,0 +1,46 @@
+    bool: {
+      filter: children.map((child) => {
+        if (child.type === 'literal') {
+          child = nodeTypes.function.buildNode('is', '*', child.value);
```
I don't think so, that would alter the AST built from query string provided by the user. Then when we call `toKueryExpression` on this node we won't know if the original query string contained an explicit `is` function, or just a literal.

In my experience so far, it seems best to read the query string as-is with minimal "interpretations" at parse time and leave things like interpreting the meaning of a literal argument to the individual serializers.

weltenwort replied:
Hm, or we could introduce an "implicit" style for `is`, which would preserve that formatting while still having an unambiguous AST. I don't have enough experience with building languages to judge the pros and cons of either solution. My intuition is that we should either create a fully "interpreted" AST as early as possible or as late as possible, but not in between.

Bargs replied:
> I don't have enough experience with building languages to judge the pros and cons of either solution. 

Nor do I ¯\\_(ツ)_/¯

Usually when faced with such uncertainty I'll just move ahead with whatever works and assume time will reveal a better solution. Let's leave as-is for now if you're ok with it.

weltenwort replied:
agreed, just something to keep in mind

Bargs commented on src/ui/public/kuery/lib/functions/geo_bounding_box.js:
```diff
@@ -0,0 +1,40 @@
+export function buildNodeParams(fieldName, params) {
+  params = _.pick(params, 'topLeft', 'bottomRight');
+  const fieldNameArg = nodeTypes.literal.buildNode(fieldName);
+  const args = _.pairs(params).map((argument) => {
```
👍  a54274a5bfe1cfa24b6f2200d1435f6db8073341

Bargs commented on src/ui/public/kuery/lib/functions/geo_bounding_box.js:
```diff
@@ -0,0 +1,40 @@
+  const field = indexPattern.fields.byName[fieldName];
+  const queryParams = args.reduce((acc, arg) => {
+    const snakeArgName = _.snakeCase(arg.name);
+    acc[snakeArgName] = ast.toElasticsearchQuery(arg);
```
👍  6d8e23f9046ddf3013f1e7f085f8726556c96706

Bargs commented on src/ui/public/kuery/lib/functions/geo_bounding_box.js:
```diff
@@ -0,0 +1,40 @@
+  const queryParams = args.reduce((acc, arg) => {
+    const snakeArgName = _.snakeCase(arg.name);
+    acc[snakeArgName] = ast.toElasticsearchQuery(arg);
+    return acc;
```
Seems a little overzealous to me, but sure :)

👍  af95b5d6eb26c218083a005a10d387e36c5e6652

Bargs commented:
@weltenwort I believe I've updated or responded to all of your inline feedback. A couple of things you mentioned in your summary:

> Another inconsistency that slightly bugs me is the varying use of literal and named arguments in the AST. Is there something keeping us from always using named arguments? I expect that would make maintaining backwards-compatibility easier when we expand the language if the order-dependent logic is restricted to only the topmost parsing layer.

Are you saying we shouldn't support unnamed parameters in the language, or that unnamed parameters in a query string should still turn into a `namedArg` node in the AST? 

If the former: Less verbosity and more fluid typing are the main arguments for unnamed parameters, I think. `is(foo, bar)` is a lot easier to type than `is(fieldName=foo, value=bar)`

If the latter: It's a possibility I've considered. I suppose we could give `namedArg` nodes a `serializeStyle` like functions have. It just doesn't seem terribly necessary at the moment. I figured I'd play around with the idea in a future refactoring.

> In the individual toKueryExpression functions there seems to be a lot of duplication. And asymmetric looking code. Maybe some kind of mapping from serialization style to a formatting function could make that more readable. But that's obviously a nice-to-have that can be delayed.

Agreed, it would be nice to DRY up these functions in a future refactoring after the syntax has seen some use and we're sure this is the way we want to go.

Thanks for all the feedback so far, keep it coming!

weltenwort commented:
> Are you saying we shouldn't support unnamed parameters in the language, or that unnamed parameters in a query string should still turn into a namedArg node in the AST?

I mean the latter. That way the interpretation of the position can happen early and only once. The idea of having a single "argument" AST node type with a style property sounds good to me, probably worth trying out in a subsequent PR.

Bargs commented:
@lukasolson thanks for the feedback, I've made some updates:

> When your search doesn't return results, if you have Kuery selected, the message still says the message about the query bar using Lucene syntax with examples.

Good catch. For now I've simply updated the no-results page to hide the bit about the lucene query syntax if kuery is selected in the query bar. Eventually I do think we'll want to fill this out with some details about kuery.

> When I create a visualization and use the filters bucket aggregation, I get errors.

This was due to my changes to `from_user.js`. I didn't realize the filter agg was using our `parse-query` directive. I've updated the filter agg's serializer to correctly wrap lucene query strings in json just like we do when building a query from the query bar.

> Not sure if this is just me, but seems like we're missing the outline on the drop-down here

This seems to be part of the UI framework (see screenshots from the framework docs below). If it's a mistake (cc @cjcenizal) we should probably fix it in a separate PR.

![screen shot 2017-07-12 at 10 06 19 am](https://user-images.githubusercontent.com/6239176/28125334-4f260302-66f4-11e7-9338-b5e582d8a01e.png)

I believe this addresses all of your points. Please take another look!

Bargs commented:
I had to do a squash in order to rebase and resolve conflicts with master. In case you guys were looking at my most recent commits incrementally I've pushed a snapshot of them [here](https://github.com/Bargs/kibana/tree/enhanceKuery.bak2).

Bargs commented:
@weltenwort @lukasolson I've finished writing tests. Now I'm just awaiting more review feedback or LGTMs.

Bargs commented on src/ui/public/filter_bar/lib/__tests__/map_geo_polygon.js:
```diff
@@ -0,0 +1,82 @@
+
+    it('should return undefined for none matching', function (done) {
+      const filter = { meta: { index: 'logstash-*' }, query: { query_string: { query: 'foo:bar' } } };
+      mapGeoPolygon(filter).catch(function (result) {
```
Honestly, this whole file is a copy/paste job from the other mapper tests. I pretty much never use `done` myself. While I think your suggestion is a good one, I think it would make more sense to apply this as a refactoring to all of the mapper tests in a separate PR if we think it's a worthwhile change.

weltenwort replied:
agreed, I did not look at the other tests

Bargs commented on src/ui/public/courier/data_source/build_query/__tests__/from_filters.js:
```diff
@@ -0,0 +1,107 @@
+    }));
+
+    it('should return the parameters of an Elasticsearch bool query', function () {
+      const result = buildQueryFromFilters();
```
👍  88744cc352e46e160dd9b02f7439235279e7a3e0

Bargs commented:
@lukasolson I made some updates and also have some follow up questions:

> Not sure exactly what's happening here but I'm seeing a bug, and here are some steps to reliably reproduce it

The linked search view was busted due to a couple of out of date variable references (post vis refactor). I think 8f05918ed91e1a7d58a81be507c978e112a66c86 fixes everything.

> range(amount, gte=-100, lt=100)

This is a bit of a conundrum. `-100` can be interpreted as both a number and a query. The query_string query has this issue as well, which it solves by forcing you to escape the `-` with a `\`. I can think of a few options for handling this:

- Copy the query_string query and require an escape. I’m not really a fan of this, requiring a `\` before every negative number is pretty unintuitive.
- Change the negation operator to something that’s not a valid part of a number. We could use `!` instead or `-`. `-` is particularly problematic because we can’t simply wrap the value in double quotes to escape it, because it would then be parsed as a string instead of a number. 
- Get rid of the implicit “all fields” query. In other words, remove the ambiguity. -100 is always a number. To execute an all fields query you would have to do `-*:100` or `-is(*, 100)`. The downside is that a new user needs to understand the query language, they can’t just start typing in search terms google style.
- Parse -100 differently depending on context. This will make things a lot more complex (both for the user and for us) and I’d rather avoid it unless we run into a situation where it’s absolutely necessary. But it’s an option.

> When I make a mistake in the syntax on Discover, I get an error, but I can keep working. However, when I make a syntax error on Visualize, at first it seems like I get an error, but then when I try to fix my error, I get a fatal error

This appears to be an issue in Visualize. I can reproduce the same error in master. @thomasneirynck @ppisljar this seems like a pretty bad bug, potentially a new blocker if it hasn't already been reported.

> Not really sure what should be done about this (if anything), but when I use the filters aggregation in Visualize, but have kuery selected, when I click on one of the bars, nothing happens. 

Yeah I mentioned this as a caveat in the PR description. I was planning on punting on it for now.

> We may need to update toUser to not display the query when it's a match all. This shows up when I select the filters aggregation, don't type anything in, then hit "play"

Ah dang, this is because the agg_type’s write method is mutating the view state. I knew I should clone the filter.input object but I was afraid something might already be relying on this mutation. I’ve updated the serializer to clone the filter input before modifying it, seems to solve this problem without creating any new ones. 6f0380a168136240dce6cc01fba5e624d8873604

> When I have kuery selected in Discover, and click on one of the bars in the date histogram, nothing happens. Ironically, when I brush, the filter works correctly.

Hah, the inconsistency with which we handle time filters is kind of hilarious. As far as I can tell, the “listeners” option that Discover was passing to `new Vis` is no longer valid after the refactor, which explains why this worked when I tested it before rebasing, but not now. The filter_bar_click_handler and brushEvent handler now get bound by the Vis constructor automatically. The filter_bar_click_handler assigns filters to the $newFilters scope variable, which the filter_bar directive picks up and handles. The brushEvent handler does the same thing with one important difference: it detects time filters and updates the global time filter itself. That’s why brushes were working, but clicks weren’t. So I just had to add the same watcher on $newFilters to Discover that I added to Visualize and Dashboard. 515a53bc33c2184936c1cf9969ca8b466eaa13fb

> Looks like TSVB is broken

Ah, this should be fun. I’ll have to check that out tomorrow.


@weltenwort

> How about extending the tests to cover the new params property of the map_* functions' return value?

I thought about this before, but it seemed fairly low value since the tests are already incomplete and we want to delete this code anyway. We didn’t bother updating them when we added `type` to the return value in the filter editor PR. But I don’t mind updating them if you believe it is worthwhile.

> When reading I sometimes was confused by finding index.js files in __tests__ directories. Do we maybe want the filename to reflect the test subject's name for easier identification?

b0d54aa449d9398a91f4407966bc4e4bd44a46fe

weltenwort commented:
> Parse -100 differently depending on context. This will make things a lot more complex (both for the user and for us) and I’d rather avoid it unless we run into a situation where it’s absolutely necessary. But it’s an option.

I think this is the only appropriate solution here unless we want to allow sub-queries (which doesn't make sense right now as we have no way to apply a projection to the query results in the language). As a user you expect the language to do "the right thing" depending on the context.

We would have to tighten up the sub-expression types anyway in order to provide helpful error messages. I think that is one of the bigger problems with the lucene query language: it is way to lenient and doesn't tell the user that his mental model is wrong. Helpful compiler error messages are (rightly so IMHO) the reason why so many relatively new languages can compete with the incumbents (see rust, elm, go, typescript, ...).

Bargs commented:
@weltenwort I'm surprised you're not in favor of removing the ambiguity entirely. After sleeping on it, it seems to me that's the real issue that needs to be solved. Parsing differently depending on context may help in *some* contexts, but not all. 

Consider the query string `-100`. Is this a search for "negative one hundred" or "not one hundred"? Context doesn't help us here. The first three options I outlined are the only ones that remove the ambiguity. Using a different negation operator seems like the most pragmatic to me so far. 

I was curious how others handle it, since most of [Lukas's examples](https://github.com/elastic/kibana/issues/12282#issuecomment-308831510) use `-`. The answer is: not well. As mentioned, the query_string query forces you to escape the `-`. Google doesn't handle negative numbers [at all](https://webapps.stackexchange.com/questions/50092/how-can-i-search-google-for-a-negative-number). Github has a weird hybrid approach that allows you to use both `NOT` and `-`, but for some reason `NOT` doesn't work with numbers. Slack doesn't seem to do negation at all. So in picking `-` as our negation operator, I think we're being led astray by bad precedents that are trying to solve different problems than we are.

Bargs commented:
@lukasolson so fixing TSVB is a bit complicated. It "integrates" with dashboard by simply [grabbing the raw ES queries](https://github.com/elastic/kibana/commit/290840509d75c8d0ac165c8029fbdb419336b8d3#diff-822e30f420592515584cb6382b5749db) from the filters and query stored in appState. With my updates the query object stored in app state is no longer a valid ES query, it is just a string that gets transformed into an ES query a request time. So my initial thought was to do the same transformation in the above file and return the raw query to TSVB, since that's what it wants. But there's a hitch, TSVB visualizations aren't associated with a Kibana index pattern. The kuery to ES methods expect an index pattern in order to support scripted fields. So until TSVB supports index patterns and ideally uses a SearchSource, it's pretty much incompatible with kuery. 

For now I've added a workaround (004cf2e2dee04372a129a8cd77a60423612d64ac) so that TSVB (and timelion as well) ignore kuery queries and continue to work with lucene queries. We can add a note about it to the docs and work toward a better solution in the future. I think that should be ok considering we're talking about interactions between two experimental features.

Let me know if that sounds reasonable to you.

Bargs commented:
@weltenwort @lukasolson barring additional feedback or responses to my responses, I think I've taken care of everything except for the negative range issue, which I'm still looking for more feedback on.

Bargs commented:
I went ahead and swapped out `-` for `!` so you can get a feel for how it would work.

Bargs commented:
> Have you considered not supporting scripted fields in kuery either? Maybe an inline "painless expression" in the language could substitute for that or even increase flexibility.

That's definitely an idea I want to play around with. Supporting scripted fields as they exist today was important to me though. I've seen a lot of users try to type their scripted field names into the query bar and then get confused when it doesn't work. Filters created via clicks on visualizations built on scripted fields also wouldn't work. Storing scripts at the index pattern level also makes it possible to tie together heterogeneous data sets. If two index patterns have scripted fields with the same name, that produce the same type of values (despite the scripts themselves being different), we can easily query that scripted field across both index patterns on a dashboard.

There's a lot of potential for improving the scripting experience though. Scripted fields are so clunky today, I'd love to transform them into something that works more seamlessly with search.

Bargs commented:
@lukasolson I fixed the issues with the range filter and the filters agg label. `or` and `and` are both case insensitive. Can you share the query you're having trouble with? Perhaps there's something else odd going on.

Bargs commented:
Test failure is unrelated to this PR

lukasolson commented:
> Can you share the query you're having trouble with? Perhaps there's something else odd going on.

It's a simple OR with conditions on either side:

![jul-20-2017 10-02-57](https://user-images.githubusercontent.com/1178348/28429746-a8fb0a64-6d32-11e7-8207-4d0ccc0a9c48.gif)

Bargs commented:
@lukasolson ah yep, I see what's wrong, looks like I was deceiving myself into thinking it was working 😅

Should be fixed now with my latest commit

Bargs commented:
@lukasolson I've run into that before, it's because of how the silly $newFilters scope variable is used. When I tried to fix it in the past I just ended up creating an infinite digest loop. I just gave it another shot and I think I got it this time.

Bargs commented:
The test that failed passed locally.... probably a timing issue

jenkins, test this

ppisljar commented:
:tada:

jimgoodwin commented:
@Bargs Please update the description above to have at least one high quality screen shot and a Release Note: paragraph at the end with a user readable summary.
